# Data Project: SQL Database 
This file is used to clean the fitbit_fitness_tracker_RAW.csv file

# Importing Libraries

In [2]:
import pandas as pd
import numpy as np
import re
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from IPython.display import display
warnings.filterwarnings('ignore')

# Loading the Fitbit CSV Data 

In [12]:
# Definig the file path 
file_path = "Datasets_files/fitbit_fitness_tracker_RAW.csv"

# Loading the CSV file into a dataframe 
fitbit_df = pd.read_csv(file_path)

# Displaying the first 10 rows of the dataframe
fitbit_df.head(10)

Id               SleepDay  TotalSleepRecords  TotalMinutesAsleep  \
0  1503960366  4/12/2016 12:00:00 AM                  1                 327   
1  1503960366  4/12/2016 12:00:00 AM                  1                 327   
2  1503960366  4/12/2016 12:00:00 AM                  1                 327   
3  1503960366  4/12/2016 12:00:00 AM                  1                 327   
4  1503960366  4/12/2016 12:00:00 AM                  1                 327   
5  1503960366  4/12/2016 12:00:00 AM                  1                 327   
6  1503960366  4/12/2016 12:00:00 AM                  1                 327   
7  1503960366  4/12/2016 12:00:00 AM                  1                 327   
8  1503960366  4/12/2016 12:00:00 AM                  1                 327   
9  1503960366  4/12/2016 12:00:00 AM                  1                 327   

   TotalTimeInBed ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0             346     5/7/2016       11992           7.71             7.71   
1             346     5/6/2016       12159           8.03             8.03   
2             346     5/1/2016       10602           6.81             6.81   
3             346    4/30/2016       14673           9.25             9.25   
4             346    4/12/2016       13162           8.50             8.50   
5             346    4/13/2016       10735           6.97             6.97   
6             346    4/25/2016       15355           9.80             9.80   
7             346     5/5/2016       14070           8.90             8.90   
8             346    4/28/2016       13154           8.53             8.53   
9             346    4/29/2016       11181           7.15             7.15   

   LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                       0.0                2.46                      2.12   
1                       0.0                1.97                      0.25   
2                       0.0                2.29                      1.60   
3                       0.0                3.56                      1.42   
4                       0.0                1.88                      0.55   
5                       0.0                1.57                      0.69   
6                       0.0                5.29                      0.57   
7                       0.0                2.92                      1.08   
8                       0.0                3.54                      1.16   
9                       0.0                1.06                      0.50   

   LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                 3.13                      0.0                 37   
1                 5.81                      0.0                 24   
2                 2.92                      0.0                 33   
3                 4.27                      0.0                 52   
4                 6.06                      0.0                 25   
5                 4.71                      0.0                 21   
6                 3.94                      0.0                 73   
7                 4.88                      0.0                 45   
8                 3.79                      0.0                 48   
9                 5.58                      0.0                 16   

   FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                   46                   175               833      1821  
1                    6                   289               754      1896  
2                   35                   246               730      1820  
3                   34                   217               712      1947  
4                   13                   328               728      1985  
5                   19                   217               776      1797  
6                   14                   216               814      2013  
7                   24                   250               857 

In [ ]:
# Reviewing the names of the columns in a list form

#print(fitbit_df.columns.tolist())
for i, col in enumerate(fitbit_df.columns):
    print(f"{i}: {col}")

0: Id
1: SleepDay
2: TotalSleepRecords
3: TotalMinutesAsleep
4: TotalTimeInBed
5: ActivityDate
6: TotalSteps
7: TotalDistance
8: TrackerDistance
9: LoggedActivitiesDistance
10: VeryActiveDistance
11: ModeratelyActiveDistance
12: LightActiveDistance
13: SedentaryActiveDistance
14: VeryActiveMinutes
15: FairlyActiveMinutes
16: LightlyActiveMinutes
17: SedentaryMinutes
18: Calories


# Reviewing the dataset for any null, missing values, and duplicates.

In [ ]:
# Obtaining and Displaying the missing data points per column 
missing_values_count = fitbit_df.isnull().sum()
missing_values_count[:]

Id                          0
SleepDay                    0
TotalSleepRecords           0
TotalMinutesAsleep          0
TotalTimeInBed              0
ActivityDate                0
TotalSteps                  0
TotalDistance               0
TrackerDistance             0
LoggedActivitiesDistance    0
VeryActiveDistance          0
ModeratelyActiveDistance    0
LightActiveDistance         0
SedentaryActiveDistance     0
VeryActiveMinutes           0
FairlyActiveMinutes         0
LightlyActiveMinutes        0
SedentaryMinutes            0
Calories                    0
dtype: int64

In [ ]:
# Checking for duplicates 
fitbit_df.duplicated().sum()

93

In [10]:
# Showing where the duplicates are
fitbit_df[fitbit_df.duplicated()]

Id               SleepDay  TotalSleepRecords  \
4764   4388161847   5/5/2016 12:00:00 AM                  1   
4765   4388161847   5/5/2016 12:00:00 AM                  1   
4766   4388161847   5/5/2016 12:00:00 AM                  1   
4767   4388161847   5/5/2016 12:00:00 AM                  1   
4768   4388161847   5/5/2016 12:00:00 AM                  1   
...           ...                    ...                ...   
11474  8378563200  4/25/2016 12:00:00 AM                  1   
11475  8378563200  4/25/2016 12:00:00 AM                  1   
11476  8378563200  4/25/2016 12:00:00 AM                  1   
11477  8378563200  4/25/2016 12:00:00 AM                  1   
11478  8378563200  4/25/2016 12:00:00 AM                  1   

       TotalMinutesAsleep  TotalTimeInBed ActivityDate  TotalSteps  \
4764                  471             495    4/17/2016        4660   
4765                  471             495    4/18/2016       11009   
4766                  471             495    4/19/2016       10181   
4767                  471             495    4/21/2016       10055   
4768                  471             495    4/15/2016        8758   
...                   ...             ...          ...         ...   
11474                 388             402     5/3/2016        8712   
11475                 388             402    4/25/2016       12405   
11476                 388             402    4/26/2016       16208   
11477                 388             402    4/23/2016        5709   
11478                 388             402    4/24/2016        3703   

       TotalDistance  TrackerDistance  LoggedActivitiesDistance  \
4764            3.58             3.58                  0.000000   
4765            9.10             9.10                  0.000000   
4766            7.83             7.83                  0.000000   
4767            7.73             7.73                  0.000000   
4768            6.73             6.73                  0.000000   
...              ...              ...                       ...   
11474           6.91             6.91                  2.253081   
11475           9.84             9.84                  2.092147   
11476          12.85            12.85                  0.000000   
11477           4.53             4.53                  0.000000   
11478           2.94             2.94                  0.000000   

       VeryActiveDistance  ModeratelyActiveDistance  LightActiveDistance  \
4764                 0.00                      0.00                 3.58   
4765                 3.56                      0.40                 5.14   
4766                 1.37                      0.69                 5.77   
4767                 0.37                      0.39                 6.98   
4768                 0.00                      0.00                 6.73   
...                   ...                       ...                  ...   
11474                1.34                      1.06                 4.50   
11475                5.05                      0.87                 3.92   
11476                7.51                      0.92                 4.42   
11477                1.52                      0.52                 2.48   
11478                0.00                      0.00                 2.94   

       SedentaryActiveDistance  VeryActiveMinutes  FairlyActiveMinutes  \
4764                       0.0                  0                    0   
4765                       0.0                 27                    8   
4766                       0.0                 20                   16   
4767                       0.0                  7                   12   
4768                       0.0                  0                    0   
...                        ...                ...                  ...   
11474                      0.0                 71                   20   
11475                      0.0                117                   16   
11476                      0.0   

Finding out the basic information of the fitbit dataset: 
- no. of rows and columns
- name of columns 
- type of values

In [15]:
# Showing the basic information of the dataset
fitbit_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12441 entries, 0 to 12440
Data columns (total 19 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Id                        12441 non-null  int64  
 1   SleepDay                  12441 non-null  object 
 2   TotalSleepRecords         12441 non-null  int64  
 3   TotalMinutesAsleep        12441 non-null  int64  
 4   TotalTimeInBed            12441 non-null  int64  
 5   ActivityDate              12441 non-null  object 
 6   TotalSteps                12441 non-null  int64  
 7   TotalDistance             12441 non-null  float64
 8   TrackerDistance           12441 non-null  float64
 9   LoggedActivitiesDistance  12441 non-null  float64
 10  VeryActiveDistance        12441 non-null  float64
 11  ModeratelyActiveDistance  12441 non-null  float64
 12  LightActiveDistance       12441 non-null  float64
 13  SedentaryActiveDistance   12441 non-null  float64
 14  VeryAc

In [ ]:
# Counting the unique ID and reviewing how much IDs it has
#fitbit_df["Id"].nunique()
unique_id = len(pd.unique(fitbit_df["Id"])) 
print("Amount of unique Id: " + str(unique_id))

# of unique Id: 24


In [ ]:
# Checking for thhe amount of rows & columns
#print(len(fitbit_df))
print(fitbit_df.shape)

12441
(12441, 19)


From the observation I conclude the following: 
1. There are no missing values and Null. 
2. 19 columns and 12441 rows.
3. ActivityDate needs to be converted from object to datetime64 type.
4. there are 24 unique Id's 
5. 93 Rows of duplicated data, however the columns ActivityDate, TotalSteps, TotalDistance,	TrackerDistance, LoggedActivitiesDistance,	VeryActiveDistance,	ModeratelyActiveDistance, LightActiveDistance, SedentaryActiveDistance,	VeryActiveMinutes, FairlyActiveMinutes,LightlyActiveMinutes, SedentaryMinutes, Calorie show different dates

In [ ]:
# Converting the ActivityDate column to datatime64 dtype and format to dd-mm-yyyy
# Converting the ActivityDate column to datatime64 dtype and format to mm-dd-yyyy
fitbit_df["ActivityDate"] = pd.to_datetime(fitbit_df["ActivityDate"], format="%m/%d/%Y").dt.date #add .dt.date to remove the time [ns]

fitbit_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12441 entries, 0 to 12440
Data columns (total 19 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Id                        12441 non-null  int64  
 1   SleepDay                  12441 non-null  object 
 2   TotalSleepRecords         12441 non-null  int64  
 3   TotalMinutesAsleep        12441 non-null  int64  
 4   TotalTimeInBed            12441 non-null  int64  
 5   ActivityDate              12441 non-null  object 
 6   TotalSteps                12441 non-null  int64  
 7   TotalDistance             12441 non-null  float64
 8   TrackerDistance           12441 non-null  float64
 9   LoggedActivitiesDistance  12441 non-null  float64
 10  VeryActiveDistance        12441 non-null  float64
 11  ModeratelyActiveDistance  12441 non-null  float64
 12  LightActiveDistance       12441 non-null  float64
 13  SedentaryActiveDistance   12441 non-null  float64
 14  VeryAc